File: cs101/pa1/Lex.java
//Lex.java
//Camille Rogers
//carogers
//Pa1
//checks there are two command line arguments, reads lines from file and puts into 
//string array, creates a List using elements from List.java to sort elements in string
//array

import java.io.*;
import java.util.Scanner;
   
class Lex{
	
   public static void main(String[] args) throws IOException{
   		Scanner in = null;			
      	PrintWriter out = null;
      	String line = null;
      	String[] array = null;		//create String array
      	int lineNumber = 0;			
      	List indices = new List();	//create List indices
      	int l = 0;
      	String current = null;
	  	String back = null;
	  	String front = null;

      	if(args.length < 2){		//check if there are two command line arguments
        	System.err.println("Usage: Lex infile outfile");
         	System.exit(1);
      	}
      
      	in = new Scanner(new File(args[0]));	//take in argument
      
      	while(in.hasNextLine()){				//while in has next line, add elements for array #
      		lineNumber++;
      		line = in.nextLine() + " ";    // add extra space so split works right
      	}
      	
      	in.close();							//close scanner
      		
      	array = new String[lineNumber];		//create String array, size of lineNumber 
      
      	in = new Scanner(new File(args[0]));			//open in/out scanner
      	out = new PrintWriter(new FileWriter(args[1]));
      
      	out.print("Array: " );	
      	
      	while(in.hasNextLine()){
      		line = in.nextLine() + " ";    	//add extra space so split works right
      		array[l] = line;				//add elements into array
      		out.print(array[l]);			//print String array
      		l++;
      	}
      	
     	out.println();						//close scanner
	 	in.close();
      
    	indices.append(0);					//put 0 into List indices 
     
     	//insertionSort indices
     	for(int k=1; k<array.length; k++){	
     		indices.moveBack();				//start at back element and move forward 
        	current = array[k]; 			//through array	
        
	    	for(int j=k-1; j >= 0 && current.compareTo(array[j])<0;j--){
				indices.movePrev(); 		//if current comes before array[j] move previous
	    	}
	    
	    	if(indices.index()!=-1){ 		
				indices.insertAfter(k);		//if index is not null, insert k after cursor
        	}
        
       		else{
        		indices.prepend(k);			//else insert k at the end of List
       		}
   
		}
		
		out.print("List of indices: ");
		out.println(indices);					//print List indices
		out.print("Sorted Array: ");
		indices.moveFront();					//move cursor to front of the List
		for(int i = 0; i< array.length; i++){	
			out.print(array[indices.get()]);	//print array element at the place of cursor
			indices.moveNext();					//in List, then move cursor next
	
		}
		
		out.close();							//close scanner
	}
}

   

File: OldFiles/cs101/pa2/Lex.c
//Lex.c
//Camille Rogers
//Pa1
//CMPS 101
//carogers
//Sorts array indirectly using List.c with a List of indices of array.

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include"List.h"

#define MAX_LEN 160

int main(int argc, char * argv[]){
	int n, count=0;
   	FILE *in, *out;
   	char line[MAX_LEN];
   	char* array[MAX_LEN];		//create new String array
   	List indices = newList();	//create new List indices

   // check command line for correct number of arguments
   if( argc != 3 ){
		printf("Usage: %s <input file> <output file>\n", argv[0]);
      	exit(1);
   }

   // open files for reading and writing 
   in = fopen(argv[1], "r");
   out = fopen(argv[2], "w");
   if( in==NULL ){
    	printf("Unable to open file %s for reading\n", argv[1]);
      	exit(1);
   }
   if( out==NULL ){
    	printf("Unable to open file %s for writing\n", argv[2]);
      	exit(1);
   }
   
	fprintf(out, "Array: \n");
	
    //get length of array 
    while( fgets(line, MAX_LEN, in) != NULL)  {
    	count++;				
    }
    
    //set aside memory for array
    for(int i = 0; i<=count;i++){
    	array[i] = calloc(1, sizeof(char* [MAX_LEN]) );
    }
	
	n=0;

	rewind(in); 	//sets file position to the beginning
   
   //puts lines into array
   	while( fgets(line, MAX_LEN, in) != NULL){
    	strcpy(array[n], line );
		n++;
 	}
 	
 	//print out array
 	for(int g=0; g<count; g++){
 		fprintf(out, "%s", array[g]);
 	}
	
	//put 0 into List indices 
   	append(indices,0);	

	//Insertion Sort for List indices
	for(int k=1; k<count; k++){
    	int j=k-1;						//sets J to element before k
     	moveBack(indices);				//start at back element and move forward
     	
        while(j >= 0 && strcmp(array[k],array[j])<0){
			movePrev(indices); 		//if current comes before array[j] move previous
			j--;
	    }
    	if(index(indices)!=-1){ 		
    		insertAfter(indices,k);		//if index is not null, insert k after cursor
        }
        else{
    		prepend(indices, k);			//else insert k at the end of List
       	}
	}
		
	fprintf(out, "\n");
	fprintf(out, "\n");
	fprintf(out, "List of indices: ");				
	fprintf(out, "\n");
	
	moveFront(indices);					//move to beginning of List indices
	for(int p=0; p<count; p++){			
    	fprintf(out, "%d ", get(indices)); //print out List indices
      	moveNext(indices);
		fprintf(out, "\n");
  	}

	fprintf(out, "\n");
	fprintf(out, "Sorted Array: \n");
	moveFront(indices);					//move to beginning of List indices
	for(int i = 0; i< count; i++){	
		fprintf(out,"%s ", array[get(indices)]); //print array element at the place of cursor
		moveNext(indices);					//in List, then move cursor next
		fprintf(out, "\n");
	}
	
	fclose(in);
   	fclose(out);
	
	int f=0;
	while(f<=count){	//free array
		free(array[f]);
		f++;
	}
	freeList(&indices);						//free List indices

    
}

